File: Language.Implementation.Patterns/python/ch02_basic_parsing_patterns/p02_lexer/ListLexer.py
import io
from Token import Token
from Lexer import Lexer
from Lexer import LexerError

class ListLexer(Lexer):
    T_TYPE_EOF = 1
    T_TYPE_NAME = 2
    T_TYPE_COMMA = 3
    T_TYPE_LBRACK = 4
    T_TYPE_RBRACK = 5
    tokenNames = ["n/a", "<EOF>", "NAME", "COMMA", "LBRACK", "RBRACK"]

    def __init__(self, input):
        super().__init__(input)

    def isLETTER(self):
        return 'a' <= self.c <= 'z' or 'A' <= self.c <= 'Z'

    def nextToken(self):
        while (self.c != Lexer.EOF):
            if self.c in [' ', '\t', '\n', '\r']:
                self.WS()
                continue
            elif self.c == ',':
                self.consume()
                return Token(ListLexer.T_TYPE_COMMA, ",")
            elif self.c == '[':
                self.consume()
                return Token(ListLexer.T_TYPE_LBRACK, "[")
            elif self.c == ']':
                self.consume()
                return Token(ListLexer.T_TYPE_RBRACK, "]")
            else:
                if self.isLETTER():
                    return self.NAME()
                else:
                    raise LexerError("invalid character: " + c)

        return Token(ListLexer.T_TYPE_EOF, "<EOF>")

    def NAME(self):
        buf = io.StringIO()
        buf.write(self.c)
        self.consume()
        while self.isLETTER():
            buf.write(self.c)
            buf.consume()

        return Token(ListLexer.T_TYPE_NAME, buf.getvalue())

    def WS(self):
        while self.c in [' ', '\t', '\n', '\r']:
            self.consume()

File: Language.Implementation.Patterns/cxx/ch02_basic_parsing_patterns/p02_lexer/lexer/ListLexer.cpp
#include "ListLexer.hpp"
#include <sstream>

const std::vector<std::string> ListLexer::tokenNames = {"n/a", "<EOF>", "NAME", "COMMA", "LBRACK", "RBRACK"};

std::string ListLexer::getTokenName(int tokenType) const {
    return tokenNames.at(tokenType);
}

Token ListLexer::nextToken() {
    while (c != EOF_CHAR) {
        switch (c) {
        case ' ': case '\t': case '\n': case '\r': WS(); continue;
        case ',': consume(); return Token(T_TYPE_COMMA, ","); 
        case '[': consume(); return Token(T_TYPE_LBRACK, "["); 
        case ']': consume(); return Token(T_TYPE_RBRACK, "]"); 
        default:
            if (isLETTER()) return NAME();
            throw LexerError(std::string("invalid character: ") + c);
        }
    }
    return Token(T_TYPE_EOF, "<EOF>");
}

Token ListLexer::NAME() {
    std::stringstream buf;
    do {
        buf << c;
        consume();
    } while (isLETTER());
    return Token(T_TYPE_NAME, buf.str());
}

void ListLexer::WS() {
    while (c == ' ' || c == '\t' || c == '\n' || c == '\r')
        consume();
}
